src/scalarstats.jl

```diff
@@ -260,37 +260,55 @@ variation(x) = ((m, s) = mean_and_std(x); s/m)
 realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
+
+"""
+    var_sample_mean(x[, weights::AbstractWeights]; mean=nothing)
+
+Return the estimated variance of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+"""
+
+
 """
-    sem(x)
+    sem(x[, weights::AbstractWeights]; mean=nothing)
 
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
 """
-function sem(x)
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+sem(x; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
+
+function sem(x, weights::UnitWeights; kwargs...)
+    if length(x) == length(weights)
+        T = promote(eltype(x), eltype(weights))
+        return sem(Iterators.map(T, x); kwargs...)
+    else
+        throw(ArgumentError("data and weights must be the same length"))
```

Comment:
I don't think the promotion is necessary. Similar to https://github.com/JuliaStats/StatsBase.jl/blob/4f02efa1ea5d3c14efa3bd0a554f8cb7b0903d65/src/weights.jl#L608-L612 one can just use
```suggestion
function sem(x, weights::UnitWeights; mean=nothing)
    length(x) != length(weights) && throw(DimensionMismatch("inconsistent array dimension"))
    return sem(x; mean=mean)
end
```